### PR TITLE
dialects (onnx): Initial onnx dialect support

### LINKED ISSUE #1894: add onnx.Add op

### PR DESCRIPTION
**Pull request by brnorris03**
onnx dialect initial support, as discussed on zulip. 

* Add elementwise binary operations, `onnx.Add`, `onnx.Sub`, `onnx.Mul`, `onnx.Div` (see [doc](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Add) for add semantics; the others are equivalent). 
* Add verification of broadcastable arguments
* Add lit tests

Resolves #1894

### CODE REVIEW

File: xdsl/dialects/onnx.py
Comment (superlopuh) [inline xdsl/dialects/onnx.py:31]:
```suggestion
    lhs = operand_def()
    rhs = operand_def()
    res = result_def(AnyTensorType)
```

File: tests/filecheck/dialects/onnx/onnx_ops.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/onnx/onnx_ops.mlir:4]: Could you please add tests for the other ops, also?

**Review by superlopuh — COMMENTED** [review]
Looking great, thank you!

File: xdsl/dialects/onnx.py
Comment (math-fehr) [inline xdsl/dialects/onnx.py:68]:
```suggestion
    assembly_format = "`(` $lhs `,` $rhs `)` attr-dict `:` `(` type($lhs) `,` type($rhs) `)` `->` type($res)"
```
Sadly I realized that I never implemented the space directive \` \` so the `:` will miss the space before.
I'm adding a fix for it right now so you could keep the same syntax!

File: tests/filecheck/dialects/onnx/onnx_ops.mlir
Comment (brnorris03) [reply tests/filecheck/dialects/onnx/onnx_ops.mlir]: Very useful, thanks! Updated

File: tests/filecheck/dialects/onnx/onnx_ops.mlir
Comment (brnorris03) [reply tests/filecheck/dialects/onnx/onnx_ops.mlir]: Done

File: xdsl/dialects/onnx.py
Comment (brnorris03) [reply xdsl/dialects/onnx.py]: Good to know, thanks. I actually wasn't using this so removed for now but will use the typevars as needed

File: xdsl/dialects/onnx.py
Comment (brnorris03) [reply xdsl/dialects/onnx.py]: I'm curious -- when do you prefer not to have type hints?

File: xdsl/dialects/onnx.py
Comment (brnorris03) [reply xdsl/dialects/onnx.py]: Ah, but it's great you support assembly_format, so much more concise, thanks!

**Review by brnorris03 — COMMENTED** [review]
Thanks for the helpful suggestions! I added some minimal validation and more tests.

File: xdsl/dialects/onnx.py
Comment (brnorris03) [inline xdsl/dialects/onnx.py:44]: Is there a better way to get the shapes of tensor types (I hope so)?

Comment (superlopuh) [reply xdsl/dialects/onnx.py:44]: Unfortunately not, we have to use isinstance checks for this
```suggestion
            if not isinstance(lhs_type := self.lhs.type, ShapedType):
                raise ValueError(
                    "onnx binary operation operands must be of type ShapedType"
                )
        lhs_shape: list[int] = [x.data for x in lhs_type.shape]
```

File: xdsl/dialects/onnx.py
Comment (superlopuh) [inline xdsl/dialects/onnx.py:32]: Just spotted this, could you please remove argument? We try to keep our APIs fully typed, and then add dynamic attributes separately. Now that I think about it, adding an attributes argument might be a good idea in general, but I'd rather keep the API consistent for now. For the custom parser, we have a number of examples of adding attributes after the operation has been created in the codebase

File: xdsl/dialects/onnx.py
Comment (superlopuh) [inline xdsl/dialects/onnx.py:66]:
```suggestion
            x == y for x, y in zip(res_shape, (v.data for v in self.res.type.shape), strict=True)
```
This avoids the temporary list creation, and throws an error if the length of the iterators is not the same. BTW it's probably a good idea to add a separate length check in the verifier with a pretty error.

File: xdsl/dialects/onnx.py
Comment (superlopuh) [reply xdsl/dialects/onnx.py]: We don't really have a strict style guide. When we migrated the code over to the new API I, regrettably, wrote a sed script that added the type hints. I now think they're just a bit noisy and not super helpful. The fact that you reached for a typealias to make it more explicit that the type corresponded to an operand is probably a good signal that they're confusing. I'd recommend against them whenever we use the `_def` things, except for existing files where we already have the type hints, where it might be worth keeping them for consistency.

File: xdsl/dialects/onnx.py
Comment (superlopuh) [reply xdsl/dialects/onnx.py]: Ah nice, I forgot about that, much cleaner. Thanks @math-fehr

File: xdsl/dialects/onnx.py
Comment (math-fehr) [reply xdsl/dialects/onnx.py:44]: Actually, you should be able to just write:
`lhs_shape = lhs_type.get_shape()`

It seems that `lhs` and `rhs` are not constrained to be a `Tensor` or a `ShapedType` in their definitions, is that expected?
Actually, are all operands and results supposed to be the same type? Here is an example of how we do it in xDSL: https://github.com/xdslproject/xdsl/blob/2ccde61cb96a8e8ea14c1313ebd6e5df514d84c0/xdsl/dialects/arith.py#L173

Comment (brnorris03) [reply xdsl/dialects/onnx.py:44]: The `lhs` and `rhs` do not have to be the same type, however, they should have the same element type if they are tensors (e.g., all have to have `int32` elements, or be an `int32` scalar) and the shapes must be broadcastable (what the current validation checks). 

I will add a check for matching element types, which is easy enough; but I don't know how to create them in a way that allows scalar or any shaped type (including tensor).

Comment (math-fehr) [reply xdsl/dialects/onnx.py:44]: Here is an example on how you could make this work:

```
class BinaryOperation(IRDLOperation):
  T = ConstraintVar("T", Attribute)
  lhs = operand_def(T | VectorOrTensorOf[T])
  rhs = operand_def(T | VectorOrTensorOf[T])
  res = result_def(T | VectorOrTensorOf[T])
```
This will ensure that `lhs`, `rhs`, and `res` are either a `Vector`, a ranked or unranked `Tensor`, and that they all have the same element type.

Making this work for `ShapedType` is a bit more complex, because we are not handling `ShapedType` well currently :/

Comment (brnorris03) [reply xdsl/dialects/onnx.py:44]: The onnx spec itself actually doesn't allow scalars (I double checked), just ranked tensors, so perhaps something like this is more appropriate I think (similar to `arith`), can you confirm?
```python
_T = TypeVar("_T", bound=Attribute)

class ElementwiseBinOpBase(IRDLOperation):
    """Base class for element-wise binary operations on tensors with Numpy-style broadcasting."""

    T = Annotated[Attribute, ConstraintVar("T"), _T]
    lhs = operand_def(TensorType[T])
    rhs = operand_def(TensorType[T])
    res = result_def(TensorType[T])
  ```
  
In `verify_` the type of `lhs.type` is `xdsl.dialects.builtin.TensorType`, however vscode (and pyright) are not able to statically determine that, and so I still get errors when trying to access and use the shapes. Not sure how to fix those, type hints (e.g., `lhs_shape: tuple[int] = self.lhs.type.get_shape()`) are not enough. So, how can I get the shapes of the operands and result without pyright errors?

Comment (superlopuh) [reply xdsl/dialects/onnx.py:44]: I'm afraid you still need to add the snippet I suggested above, or something like it

Comment (brnorris03) [reply xdsl/dialects/onnx.py:44]: I did add the snippet above (and variations) but cannot get types resolved statically, sorry if this is basic, but am kind of out of ideas of what to try to make pyright happy. I can push the latest if it helps.

Comment (brnorris03) [reply xdsl/dialects/onnx.py]: Actually, never mind, I had a silly mistake. I was able to get rid of the pyright errors.

Comment (math-fehr) [reply xdsl/dialects/onnx.py]: Yes, your snippet seems correct!
The main issue here is that you cannot really statically guarantee that `self.lhs` is a `TensorType`, because if you are in the middle of a rewrite, then you cannot guarantee that your operation verifies.
I'm not completely sure how to fix that currently, to be honest, MLIR has now the concept of `TypedValue` that we could use, but I don't have an easy way of integrating that yet without making it a hassle to use :/

File: xdsl/dialects/onnx.py
Comment (math-fehr) [inline xdsl/dialects/onnx.py:22]: I think you can remove this now?

File: xdsl/dialects/onnx.py
Comment (math-fehr) [inline xdsl/dialects/onnx.py:28]:
```suggestion
    T = Annotated[Attribute, ConstraintVar("T")]
```
Since you don't use `_T` in `ElementwiseBinOpBase`, you don't really need this.

File: xdsl/dialects/onnx.py
Comment (math-fehr) [inline xdsl/dialects/onnx.py:25]:
```suggestion
class ElementwiseBinOpBase(IRDLOperation, ABC):
```
To be sure we don't make the mistake of instantiating it directly.

File: xdsl/dialects/onnx.py
Comment (math-fehr) [inline xdsl/dialects/onnx.py:42]:
```suggestion
    def __init__(
        self,
        lhs: SSAValue,
        rhs: SSAValue,
        res_type: Attribute | None
    ):
        if res_type is None:
          res_type = lhs.type
        super().__init__(
            operands=[lhs, rhs],
            result_types=[res_type],
        )
```
The `res_type` argument is quite useful in the middle of a rewrite, when the operation would not verify.
Making it optional is quite useful as well, if by default you take the type of the first operand by example.

File: xdsl/dialects/onnx.py
Comment (math-fehr) [inline xdsl/dialects/onnx.py:56]: I would make an `assert` for this here, as the `verify_` function is always called after the IRDL verifier, which will always ensure that these are `TensorType`.
The `assert` makes it clearer that this is for pyright only.

**Review by math-fehr — APPROVED** [review]
Nice!
I just made some small style comments, but otherwise that looks good to me!
I would just wait for Sasha's approval.

File: xdsl/dialects/onnx.py
Comment (superlopuh) [inline xdsl/dialects/onnx.py:105]: nit:
```suggestion
ONNX = Dialect("onnx", [Add, Sub, Mul, Div,])
```
Adding a comma here will make the formatter move the ops to their own lines, which will minimise diffs in the future. A really minor nit, feel free to ignore :)

File: xdsl/dialects/onnx.py
Comment (brnorris03) [reply xdsl/dialects/onnx.py:42]: Makes sense (my initial attempt was based on examples, which may not have been the most appropriate). I think by default the result type should be computed based on both operands, but in any case, the suggested modification is better.

Comment (brnorris03) [reply xdsl/dialects/onnx.py:42]: I think intiially, at least, I would not include `None` as an option to the result type because I think it's not essential to be able to omit.

File: xdsl/dialects/onnx.py
Comment (brnorris03) [reply xdsl/dialects/onnx.py:28]: Related to this, I tried to figure out how to specify that the element type should be any of a specific subset of builtin types -- is there a good way to do this?

File: xdsl/dialects/onnx.py
Comment (math-fehr) [reply xdsl/dialects/onnx.py]: Sure, works as well!

File: xdsl/dialects/onnx.py
Comment (math-fehr) [reply xdsl/dialects/onnx.py]: Yes, just replace the `Attribute` with the type you want, something like:
```
T = Annotated[Float32Type | IntegerType, ConstraintVar("T")]
```

Comment (brnorris03) [reply xdsl/dialects/onnx.py]: Thanks! Changed to `T = Annotated[AnyFloat | IntegerType, ConstraintVar("T")]`, which matches the spec more closely.

**Review by georgebisbas — APPROVED** [review]
Looks good!

**Review by superlopuh — APPROVED** [review]
Perfect, thank you!

**Pull request merged**
